File: LinkedIn/254.factor-combinations.java
import java.util.*;

/**
 * Algorithm:
 * Backtrack nondecreasing factors. For every factor that divides target, emit
 * path + factor + target/factor, then continue factoring target/factor starting
 * from the same factor.
 *
 * Complexity:
 * Output-dependent; recursion depth O(log n) in factor count.
 */
class Solution {
    private List<List<Integer>> ans;

    public List<List<Integer>> getFactors(int n) {
        ans = new ArrayList<>();
        dfs(2, n, new ArrayList<>());
        return ans;
    }

    private void dfs(int start, int target, List<Integer> path) {
        for (int factor = start; factor * factor <= target; factor++) {
            if (target % factor == 0) {
                List<Integer> combo = new ArrayList<>(path);
                combo.add(factor);
                combo.add(target / factor);
                ans.add(combo);

                path.add(factor);
                dfs(factor, target / factor, path);
                path.remove(path.size() - 1);
            }
        }
    }
}

File: data structure and algorithms/Backtracking/254.factor-combinations.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    vector<vector<int>> getFactors(int n) {
        /*
        Approach: backtrack nondecreasing factors. For each factor dividing the
        current target, output the current path plus factor and target/factor,
        then continue factoring target/factor starting at the same factor.

        Complexity: output-sensitive exponential time, O(log n) recursion depth
        for factor chains.
        */
        vector<vector<int>> ans;
        vector<int> path;
        function<void(int, int)> dfs = [&](int start, int target) {
            for (int factor = start; (long long)factor * factor <= target; ++factor) {
                if (target % factor != 0) continue;
                path.push_back(factor);
                vector<int> combo = path;
                combo.push_back(target / factor);
                ans.push_back(combo);
                dfs(factor, target / factor);
                path.pop_back();
            }
        };
        dfs(2, n);
        return ans;
    }
};
